rfcs/013-gitlab-spam-checker/013-gitlab-spam-checker.md

```diff
@@ -0,0 +1,177 @@
+# Implementing a Spam Management Service for GitLab
+
+- Start Date: 2023-05-19
+- Status: Proposed
+
+## Summary
+
+This proposal aims to introduce a spam management service to mitigate spam on GitLab and improve user experience and security. This service will use machine learning models to detect and eliminate spam across the platform in real-time where possible, focusing on snippets, issues, notes, projects, and user profiles.
+
+## Problem Statement
+
+GitLab, a service on which RenkuLab relies, is not immune to the issue of spam. Spam can infiltrate various platform aspects, including snippets, issues, notes, projects, and user profiles. Spam disrupts the user experience, poses potential security risks, and damages the reputation and RenkuLab.io's SEO. GitLab lacks a robust, real-time solution to identify and eliminate spam across all attack vectors.
+
+## Key Assumptions
+
+- The proposed machine learning model will be built in consultation with the innovation team. The machine learning model's specifics are out of this RFC's scope.
+
+## Possible Solutions
+
+- A service that only checks for spam at regular intervals. This solution is less than ideal because of the delay between the spam being posted and the spam being detected and removed, during which spam could already be harmful. 
+
+- Monitor GitLab for spam manually. This solution is unsuitable because it is time-consuming.
+
+- GitLab has its [Spamcheck](https://docs.gitlab.com/ee/user/admin_area/reporting/spamcheck.html) service, which can be used to check for spam issues and snippets. This service does not check for spam in notes, projects, or user profiles. Currently, it is not possible to configure the service to delete spam automatically. Deploying GitLab's Spamcheck service also requires GitLab Enterprise Edition, which RenkuLab.io's GitLab is not currently running on.
```

Comment:
Do we have access to the Spamcheck code? And if yes, can we deploy it and call its API ourselves? This would mean we do not have to develop the model ourselves.

Reply:
The Spamcheck code is [public](https://gitlab.com/gitlab-org/gl-security/security-engineering/security-automation/spam/spamcheck). The model, however, is not included in the repo and is instead downloaded from an S3 bucket when the service starts. I have done what you have suggested for evaluating existing content on our GitLab for spam. There are some improvements we can make, though. GitLab themselves have disabled the ability of Spamcheck to remove content from GitLab, presumably because they don't trust its results yet: https://gitlab.com/gitlab-org/gl-security/security-engineering/security-automation/spam/spamcheck/-/blob/v1.4.1/app/spammable.py#L26-33

Reply:
The main problem with Spamcheck is that it doesn't check anything besides snippets and issues. So it doesn't fulfil all of our requirements.

Reply:
but that's a question of its integration into gitlab's functionality - the model is still available to be used to validate the events as you propose, no?

Reply:
Not really - I want the model to be able to detect the presence or absence of attributes as a way of identifying spam.